### PR TITLE
add optional exclusive writes to rocksdb engine

### PR DESCRIPTION
**Pull request by ObiWahn**
- [x] add tests
- [x] add documentation
  - https://github.com/arangodb/docs/pull/220 - release notes
  - docs for startup options are generated automatically
- [x] add CHANGELOG entry

### CODE REVIEW

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 1]
```diff
@@ -342,6 +343,13 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
                   arangodb::options::makeFlags(arangodb::options::Flags::Hidden))
       .setIntroducedIn(30405)
       .setIntroducedIn(30500);
+  options
+      ->addOption("--rocksdb.exclusive-writes",
+                  "if true, writes are exclusive. This allows the RocksDB engine to mimic "
+                  "the MMfiles collection locks ",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp]:
```suggestion
                  "the MMfiles collection locks behavior, but will inhibit concurrent write operations",
```

**Comment by Simran-B** [comment]
Re docs, please add a sentence or two when it makes sense for users to use this option (e.g. single user ingesting a lot of data and running analytical workloads) and what to expect in terms of performance for the ideal and the worst-case workload.

File: arangod/RocksDBEngine/RocksDBTransactionCollection.cpp [hunk 2]
```diff
@@ -45,7 +46,10 @@ RocksDBTransactionCollection::RocksDBTransactionCollection(TransactionState* trx
       _numInserts(0),
       _numUpdates(0),
       _numRemoves(0),
-      _usageLocked(false) {}
+      _usageLocked(false),
+      _exclusiveWrites(trx->vocbase().server().getFeature<arangodb::RocksDBOptionFeature>()._exclusiveWrites)
+      {}
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: why is it necessary to have this bool in every single transaction ?

Comment (ObiWahn) [reply arangod/RocksDBEngine/RocksDBTransactionCollection.cpp]: How would you do it?

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 3]
```diff
@@ -342,6 +343,13 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
                   arangodb::options::makeFlags(arangodb::options::Flags::Hidden))
       .setIntroducedIn(30405)
       .setIntroducedIn(30500);
+  options
+      ->addOption("--rocksdb.exclusive-writes",
+                  "if true, writes are exclusive. This allows the RocksDB engine to mimic "
+                  "the MMfiles collection locks behavior, but will inhibit concurrent write operations",
+                  new BooleanParameter(&_exclusiveWrites))
+      .setIntroducedIn(30505) //TODO fix version
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp]:
```suggestion
      .setIntroducedIn(30504) 
```

File: arangod/StorageEngine/RocksDBOptionFeature.cpp [hunk 4]
```diff
@@ -342,6 +343,13 @@ void RocksDBOptionFeature::collectOptions(std::shared_ptr<ProgramOptions> option
                   arangodb::options::makeFlags(arangodb::options::Flags::Hidden))
       .setIntroducedIn(30405)
       .setIntroducedIn(30500);
+  options
+      ->addOption("--rocksdb.exclusive-writes",
+                  "if true, writes are exclusive. This allows the RocksDB engine to mimic "
+                  "the MMfiles collection locks behavior, but will inhibit concurrent write operations",
```
Comment (jsteemann) [inline arangod/StorageEngine/RocksDBOptionFeature.cpp]:
```suggestion
                  "the MMFiles collection locks behavior, but will inhibit concurrent write operations",
```

File: tests/js/client/server_permissions/exclusive-collections-noncluster-rocksdb.js [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        command: function() {
+          let db = require("internal").db;
+          db.UnitTestsExclusiveCollection2.insert({ _key: "runner1", value: false });
+
+          while (!db.UnitTestsExclusiveCollection2.exists("runner2")) {
+            require("internal").sleep(0.02);
+          }
+
+          db._executeTransaction({
+            collections: { write: [ "UnitTestsExclusiveCollection1", "UnitTestsExclusiveCollection2" ], exclusive: [ ] },
```
Comment (jsteemann) [inline tests/js/client/server_permissions/exclusive-collections-noncluster-rocksdb.js]:
```suggestion
            collections: { write: [ "UnitTestsExclusiveCollection1", "UnitTestsExclusiveCollection2" ] },
```

File: tests/js/client/server_permissions/exclusive-collections-noncluster-rocksdb.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        }
+      });
+
+      db.UnitTestsExclusiveCollection2.insert({ _key: "runner2", value: false });
+      while (!db.UnitTestsExclusiveCollection2.exists("runner1")) {
+        require("internal").sleep(0.02);
+      }
+
+      db._executeTransaction({
+        collections: { write: [ "UnitTestsExclusiveCollection1", "UnitTestsExclusiveCollection2" ], exclusive: [ ] },
```
Comment (jsteemann) [inline tests/js/client/server_permissions/exclusive-collections-noncluster-rocksdb.js]:
```suggestion
        collections: { write: [ "UnitTestsExclusiveCollection1", "UnitTestsExclusiveCollection2" ] },
```

**Comment by jsteemann** [comment]
https://172.16.10.101/view/PR/job/arangodb-matrix-pr/7122/

**Comment by ObiWahn** [comment]
tests/js/client/server_permissions/exclusive-collections-noncluster-rocksdb.js - touched - passed locally

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7131/

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/7132/

**Pull request merged**